File: Test/WrapperClipperCSharp/Wrapper.vb
Imports PclWCommon
Imports PclWTestCommon


Public Class ClipperCSharpWrapper : Inherits TestableObj
    Private Const Scale As Integer = 1000

    Public Sub New()
        Me.ClosedPolygonsRequired = False
    End Sub

    Public Overrides Function GetAdaptedInputFromPolygonSet(ByVal input As PclWCommon.PolygonSet) As Object
        If Me.ClosedPolygonsRequired Then
            ClosePolygons(input)
        Else
            OpenPolygons(input)
        End If

        Dim clprPolygons As New List(Of List(Of ClipperLib.IntPoint))
        For Each polygon As PclWCommon.Polygon In input.Polygons
            Dim clprPolygon As New List(Of ClipperLib.IntPoint)
            For Each vertex As PclWCommon.Vertex In polygon.Vertices
                Dim clprPoint As New ClipperLib.IntPoint
                clprPoint.X = vertex.X * Scale
                clprPoint.Y = vertex.Y * Scale
                clprPolygon.Add(clprPoint)
            Next
            clprPolygons.Add(clprPolygon)
        Next
        Return clprPolygons
    End Function

    Public Overrides Function GetAdaptedInputFromRegion(ByVal input As PclWCommon.Region) As Object
        'Not implemented.
        Return Nothing
    End Function

    Public Overrides Function GetAdaptedOutputToPolygonSet(ByVal output As Object) As PclWCommon.PolygonSet
        Dim polygonSet As New PclWCommon.PolygonSet
        For Each clprPolygon As List(Of ClipperLib.IntPoint) In output
            Dim polygon As New PclWCommon.Polygon
            For Each clprPoint As ClipperLib.IntPoint In clprPolygon
                Dim vertex As New PclWCommon.Vertex
                vertex.X = clprPoint.X / Scale
                vertex.Y = clprPoint.Y / Scale
                polygon.Vertices.Add(vertex)
            Next
            polygonSet.Polygons.Add(polygon)
        Next
        Return polygonSet
    End Function

    Public Overrides Function GetDifference(ByVal subject As Object, ByVal clip As Object) As Object
        Dim clipperCs As New ClipperLib.Clipper
        'clipperCs.UseFullCoordinateRange = False
        clipperCs.AddPolygons(subject, ClipperLib.PolyType.ptSubject)
        clipperCs.AddPolygons(clip, ClipperLib.PolyType.ptClip)
        Dim result As New List(Of List(Of ClipperLib.IntPoint))
        clipperCs.Execute(ClipperLib.ClipType.ctDifference, result)
        Return result
    End Function

    Public Overrides Function GetIntersection(ByVal subject As Object, ByVal clip As Object) As Object
        Dim clipperCs As New ClipperLib.Clipper
        'clipperCs.UseFullCoordinateRange = False
        clipperCs.AddPolygons(subject, ClipperLib.PolyType.ptSubject)
        clipperCs.AddPolygons(clip, ClipperLib.PolyType.ptClip)
        Dim result As New List(Of List(Of ClipperLib.IntPoint))
        clipperCs.Execute(ClipperLib.ClipType.ctIntersection, result)
        Return result
    End Function

    Public Overrides Function GetUnion(ByVal subject As Object, ByVal clip As Object) As Object
        Dim clipperCs As New ClipperLib.Clipper
        'clipperCs.UseFullCoordinateRange = False
        clipperCs.AddPolygons(subject, ClipperLib.PolyType.ptSubject)
        clipperCs.AddPolygons(clip, ClipperLib.PolyType.ptClip)
        Dim result As New List(Of List(Of ClipperLib.IntPoint))
        clipperCs.Execute(ClipperLib.ClipType.ctUnion, result)
        Return result
    End Function

    Public Overrides Function GetXor(ByVal subject As Object, ByVal clip As Object) As Object
        Dim clipperCs As New ClipperLib.Clipper
        'clipperCs.UseFullCoordinateRange = False
        clipperCs.AddPolygons(subject, ClipperLib.PolyType.ptSubject)
        clipperCs.AddPolygons(clip, ClipperLib.PolyType.ptClip)
        Dim result As New List(Of List(Of ClipperLib.IntPoint))
        clipperCs.Execute(ClipperLib.ClipType.ctXor, result)
        Return result
    End Function
End Class

File: PclWClipper/PclWClipper/Wrapper.cpp
#include "clipper.hpp"





namespace PclWClipper{

using namespace System;
using namespace System::Collections::ObjectModel;
using namespace PclWCommon;

namespace clpr = ClipperLib;
using namespace clpr;





typedef clpr::ClipType ClprClipType;
typedef clpr::IntPoint ClprPoint;
typedef clpr::Polygon ClprPolygon;
typedef clpr::Polygons ClprPolygons;
typedef clpr::Clipper Clipper;

public ref class ClipperWrapper
{
public:
    static PolygonSet^ GetDifference(PolygonSet^ subject, PolygonSet^ clip)
    {
        return Clip(subject, clip, ctDifference);
    }

    static PolygonSet^ GetIntersection(PolygonSet^ subject, PolygonSet^ clip)
    {
        return Clip(subject, clip, ctIntersection);
    }

    static PolygonSet^ GetUnion(PolygonSet^ subject, PolygonSet^ clip)
    {
        return Clip(subject, clip, ctUnion);
    }

    static PolygonSet^ GetXor(PolygonSet^ subject, PolygonSet^ clip)
    {
        return Clip(subject, clip, ctXor);
    }    

private:
    static const double SCALE = 1000;

    static PolygonSet^ Clip(PolygonSet^ subject, PolygonSet^ clip,
        ClprClipType operation)
    {
        ClprPolygons clprSubject;
        FromPolygonSetToClprPolygons(subject, clprSubject);
        ClprPolygons clprClip;
        FromPolygonSetToClprPolygons(clip, clprClip);

        Clipper clipperEngine;
        //clipperEngine.UseFullCoordinateRange(false);
        clipperEngine.AddPolygons(clprSubject, ptSubject);
        clipperEngine.AddPolygons(clprClip, ptClip);
        ClprPolygons clprResult;
        clipperEngine.Execute(operation, clprResult);
        
        PolygonSet^ result = gcnew PolygonSet;
        FromClprPolygonsToPolygonSet(clprResult, result);
        return result;
    }

    static void FromClprPolygonsToPolygonSet(
        const ClprPolygons& clprPolygons, PolygonSet^ polygonSet)
    {
        Collection<PclWCommon::Polygon^>^ polygons = polygonSet->Polygons;
        for (unsigned int i = 0; i < clprPolygons.size(); i++)
        {
            PclWCommon::Polygon^ polygon = gcnew PclWCommon::Polygon;
            FromClprPolygonToPolygon(clprPolygons[i], polygon);
            polygons->Add(polygon);
        }
    }
    
    static void FromClprPolygonToPolygon(const ClprPolygon& clprPolygon,
        PclWCommon::Polygon^ polygon)
    {
        Collection<Vertex>^ vertices = polygon->Vertices;
        for (unsigned int j = 0; j < clprPolygon.size(); j++)
        {
            ClprPoint point = clprPolygon[j];
            Double x = point.X / SCALE;
            Double y = point.Y / SCALE;
            vertices->Add(Vertex(x, y));
        }
    }

    static void FromPolygonSetToClprPolygons(PolygonSet^ polygonSet,
        ClprPolygons& clprPolygons)
    {
        Collection<PclWCommon::Polygon^>^ polygons = polygonSet->Polygons;
        clprPolygons.resize(polygons->Count);
        for (Int32 i = 0; i < polygons->Count; i++)
        {
            FromPolygonToClprPolygon(polygons->default[i], 
                clprPolygons[i]);
        }
    }
    
    static void FromPolygonToClprPolygon(PclWCommon::Polygon^ polygon, 
        ClprPolygon& clprPolygon)
    {
        Collection<Vertex>^ vertices = polygon->Vertices;
        clprPolygon.resize(vertices->Count);
        for (Int32 j = 0; j < vertices->Count; j++)
        {
            Vertex vertex = vertices->default[j];
            ClprPoint clprPoint;
            clprPoint.X = Int32(vertex.X * SCALE);
            clprPoint.Y = Int32(vertex.Y * SCALE);
            clprPolygon[j] = clprPoint;
        }
    }
};

}